File: 16-3sum-closest/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int max = Integer.MAX_VALUE;
        int min = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            if (i > 0 && nums[i] == nums[i - 1])
                continue;
            int start = i + 1;
            int end = nums.length - 1;
            while (start < end) {
                int sum = nums[i] + nums[start] + nums[end];
                if (sum > target) {
                    max = Math.min(sum, max);
                    end--;
                } else if (sum < target) {
                    min = Math.max(sum, min);
                    start++;
                } else
                    return target;
            }
        }

        if (max != Integer.MAX_VALUE && min != Integer.MIN_VALUE)
            if (Math.abs(max - target) > Math.abs(min - target))
                return min;
            else
                return max;

        if (max != Integer.MAX_VALUE)
            return max;

        return min;
    }
}

File: 16-3sum-closest/3sum-closest.go
func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func threeSumClosest(nums []int, target int) int {
	if len(nums) == 3 {
		return nums[0] + nums[1] + nums[2]
	}

	closestSum := 0
	closestDiff := 0

	madeSum := false

	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			firstSum := nums[i] + nums[j]

			for k := j + 1; k < len(nums); k++ {
				sum := firstSum + nums[k]
				diff := Abs(target - sum)

				if !madeSum || diff < closestDiff {
					closestSum = sum
					closestDiff = diff
					madeSum = true
				}

				if diff == 0 {
					return closestSum
				}
			}
		}
	}

	return closestSum
}